### PR TITLE
HTTP proxy server for Eth2 APIs

### PR DESCRIPTION
**Pull request by rkapka**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

gRPC and HTTP are not best friends. [Eth2 APIs spec](https://ethereum.github.io/eth2.0-APIs/) includes several particularities that cannot be currently achieved with gRPC, even with the help of grpc-gateway. These are:
- sending hex values to gRPC functions expecting byte arrays
- returning errors with fields other than `message` and `code`
- returning lowercase enum values
- returning timestamps as UNIX time
- sending hex values to URL parameters
- sending an array as the top-level JSON value (an array is a valid JSON) when submitting attestations
- sending less than 32 bytes to the graffiti field, which is tagged `ssz:"size=32"`

This PR implements a gorilla-mux HTTP server that acts as middleware between the client and the grpc-gateway server. It handles all the above issues by intercepting the client request, processing it as required to make gRPC happy and forwarding it to grpc-gateway. It does the same on the way back: intercept the response from grpc-gateway, process it to comply to Eth2 API spec and forward it to the client. The proxy server achieves all of this by making extensive use of structs defined in `middleware_structs.go`, which are mostly representations of proto structs, annotated specifically for the proxy server and its field processing functions.  

**Which issues(s) does this PR fix?**

Part of overall API work

**Other notes for review**

Tests will come in another PR. This PR is big enough as it is. What's more, we should review this asap.

### CODE REVIEW

File: shared/gateway/gateway.go [hunk 1]
```diff
@@ -192,6 +221,18 @@ func (g *Gateway) Start() {
 			return
 		}
 	}()
+
+	go func() {
+		proxy := &ApiProxyMiddleware{
+			GatewayAddress: g.gatewayAddr,
+			ProxyAddress:   ":4500",
```
Comment (rkapka) [inline shared/gateway/gateway.go:228]: Hard-coded for the time being. I will add a flag in a future PR.

File: beacon-chain/rpc/beaconv1/pool.go [hunk 2]
```diff
@@ -60,6 +85,19 @@ func (bs *Server) SubmitAttestations(ctx context.Context, req *ethpb.SubmitAttes
...
+	if len(attFailures) > 0 {
+		failuresContainer := &attestationsVerificationFailure{Failures: attFailures}
+		jsonFailures, err := json.Marshal(failuresContainer)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "Could not prepare attestation failure information: %v", err)
+		}
+		if err := grpc.SetHeader(ctx, metadata.Pairs(grpcutils.CustomErrorMetadataKey, string(jsonFailures))); err != nil {
+			return nil, status.Errorf(codes.Internal, "Could not prepare attestation failure information: %v", err)
+		}
+		return nil, status.Errorf(codes.Internal, "One or more attestations failed validation")
```
Comment (rauljordan) [inline beacon-chain/rpc/beaconv1/pool.go:98]: this stuff seems like a candidate for a helper function or package or middleware. It shouldn't be part of the basic logic of an endpoint

File: shared/gateway/api_middleware.go [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+type fieldProcessor struct {
+	tag string
+	f   func(value reflect.Value) error
+}
+
+// Run starts the proxy, registering all proxy endpoints on ApiProxyMiddleware.ProxyAddress.
+func (m *ApiProxyMiddleware) Run() error {
+	m.router = mux.NewRouter()
+
+	m.handleApiEndpoint("/eth/v1/beacon/genesis")
```
Comment (rauljordan) [inline shared/gateway/api_middleware.go:50]: this list of endpoints should be passed into this function. The shared/ package is meant to be generic and does not need to know anything about the world outside it

File: shared/gateway/api_middleware.go [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+	m.handleApiEndpoint("/eth/v1/debug/beacon/states/{state_id}")
+	m.handleApiEndpoint("/eth/v1/debug/beacon/heads")
+	m.handleApiEndpoint("/eth/v1/config/fork_schedule")
+	m.handleApiEndpoint("/eth/v1/config/deposit_contract")
+	m.handleApiEndpoint("/eth/v1/config/spec")
+
+	return http.ListenAndServe(m.ProxyAddress, m.router)
+}
+
+func (m *ApiProxyMiddleware) handleApiEndpoint(endpoint string) {
```
Comment (rauljordan) [inline shared/gateway/api_middleware.go:78]: this function is really big and I would imagine hard to debug. Let's split it up

File: shared/gateway/api_middleware.go [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+		if err := grpcResp.Body.Close(); err != nil {
+			e := fmt.Errorf("could not close response body: %w", err)
+			writeError(writer, &DefaultErrorJson{Message: e.Error(), Code: http.StatusInternalServerError}, nil)
+			return
+		}
+	})
+}
+
+// Posted graffiti needs to have length of 32 bytes, but client is allowed to send data of any length.
+func prepareGraffiti(data *endpointData) {
```
Comment (rauljordan) [inline shared/gateway/api_middleware.go:236]: this package shouldn't know anything about graffiti or actual API items IMO. Maybe there can be a list of callbacks specified on initialization that you can call along the way, and one of them is to handle graffiti and attestations arrays

File: shared/gateway/gateway.go (hunk 1 above)
Comment (rkapka) [reply shared/gateway/gateway.go:228]: I will update https://github.com/prysmaticlabs/prysm/issues/7510 today.

File: beacon-chain/rpc/beaconv1/pool.go (hunk 2 above)
Comment (rkapka) [reply beacon-chain/rpc/beaconv1/pool.go:98]: Agreed, I will change this.

**Review by rauljordan — APPROVED** [review]

File: shared/gateway/api_middleware.go (hunk 3 above)
Comment (rkapka) [reply shared/gateway/api_middleware.go:50]: Endpoints extracted to `/beacon-chain/rpc`.

I don't think there is any way to enforce endpoint registration. We need documentation for this. I am thinking about writing a dev wiki page.

File: shared/gateway/api_middleware.go [hunk 6]
```diff
@@ -0,0 +1,463 @@
...
+		return err
+	}
+	v.SetString(strconv.FormatUint(uint64(t.Unix()), 10))
+	return nil
+}
+
+// getEndpointData constructs and returns a struct containing necessary information to process a request based on the provided endpoint.
+// The returned struct is meant to be used during a single request.
+func getEndpointData(endpoint string) (endpointData, error) {
+	switch endpoint {
```
[the remark this answers is not in the document]
Comment (rkapka) [reply shared/gateway/api_middleware.go:409]: Extracted to `/beacon-chain/rpc`

File: shared/gateway/middleware_structs.go [hunk 7]
```diff
@@ -0,0 +1,413 @@
+package gateway
```
[the remark this answers is not in the document]
Comment (rkapka) [reply shared/gateway/middleware_structs.go:1]: Exctracted to `/beacon-chain/rpc`.

File: shared/gateway/api_middleware.go [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
+	}
+	return nil
+}
+
+// handleUrlParameters processes URL parameters, allowing parameterized URLs to be safely and correctly proxied to grpc-gateway.
+func handleUrlParameters(endpoint string, request *http.Request, writer http.ResponseWriter) {
+	segments := strings.Split(endpoint, "/")
+	for i, s := range segments {
+		// We only care about segments which are parameterized.
+		if len(s) > 0 && s[0] == '{' && s[len(s)-1] == '}' {
```
[the remark this answers is not in the document]
Comment (rkapka) [reply shared/gateway/api_middleware.go:264]: Done

File: shared/gateway/api_middleware.go (hunk 4 above)
Comment (rkapka) [reply shared/gateway/api_middleware.go:78]: I extracted several small functions.

File: shared/gateway/api_middleware.go (hunk 5 above)
Comment (rkapka) [reply shared/gateway/api_middleware.go:234]: Done exactly that.

**Pull request merged**